### PR TITLE
Add Spatial Collapsed Spectrum selection to Line Analysis

### PR DESCRIPTION
**Pull request by duytnguyendtn**
### Description

Previously, it was not possible to perform line analysis on the spectrum generated by Cubeviz' autocollapse for spatial subsets. This PR adds that feature. 

BIG THANKS to @pllim for fixing the mysterious test failure!

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py [hunk 1]
```diff
@@ -97,6 +100,15 @@ def __init__(self, *args, **kwargs):
         # require entries to be in spectrum-viewer (not other cubeviz images, etc)
         self.dataset.add_filter('layer_in_spectrum_viewer')
 
+        if self.app.state.settings.get("configuration") == "cubeviz":
+            self.spatial_subset = SubsetSelect(self,
+                                               'spatial_subset_items',
+                                               'collapsed_spectrum_selected',
+                                               default_text='Entire Cube Spectrum',
+                                               allowed_type='spatial')
```
Comment (kecnry) [inline jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:108]: I'd advocate for using the same traitlet names and default text as in model fitting so there is a consistent UX and API for the user between the two.  If there are subtle differences that we want to capture in the UI, then the hint might be the easiest place to make that distinction.  (Or we can always change model fitting to match if there is a good reason to make changes).

File: jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py [hunk 2]
```diff
@@ -41,6 +44,50 @@ def test_plugin(specviz_helper, spectrum1d):
...
+    cubeviz_helper.app.get_viewer('flux-viewer').session.edit_subset_mode._mode = NewMode
+    # The cube only has one slice, so we have to grab the entire viewer's range to create a region
+    cubeviz_helper.app.get_viewer('spectrum-viewer').apply_roi(
+        XRangeROI(cubeviz_helper.app.get_viewer('spectrum-viewer').state.x_min,
+                  cubeviz_helper.app.get_viewer('spectrum-viewer').state.x_max))
+
+    cubeviz_helper.app.state.drawer = True
+
+    plugin = cubeviz_helper.app.get_tray_item_from_name('specviz-line-analysis')
+    plugin.open_in_tray()
```
Comment (duytnguyendtn) [inline jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:77]: Ok I've hammered this PR as far as I could, and now I have to concede I'm stuck and have no idea why the plugin is refusing to give this test its values

So currently, the tests are failing because the plugin results are still nans at the check. I can't determine why this is; when I copy and paste this entire test into a jupyter notebook, it returns the results as expected.

Slightly related is the only way I how to actually trigger the recalculation of the results via the API is by opening the tray, as the other tests in this suite do as well. As I understand it, this line is supposed to force the plugin to calculate the results, but it's not working. I've also tried calling `plugin._calculate_statistic` manually, but doesn't seem to work; it returns nans if the tray isn't opened. If the tray is open, it doesn't change the behavior.

The only thing I can think of is that there's a race condition between when the fields are being set and when the results are being read. But even that I find unlikely considering the results are supposed to be calculated for the default values immediately when the tray is opened.

Hopefully someone can point out where the mistake I'm making is. Please feel free to commandeer this PR if you find it if I can't get to it!

Comment (pllim) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:77]: If you can reproduce this locally, I would suggest you copy paste the test case into a notebook and debug there, if you have not already. If you can reproduce this failure that way, you can then drill down on the individual elements and so on.

Or is this the one you said you cannot reproduce locally and only on CI? I cannot keep track of all the conversations.

Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:77]: > is this the one you said you cannot reproduce locally and only on CI?

Yep, this is the one... My local pytest runs give me the green light. @kecnry graciously tested it and saw the same behavior as I. I suppose the next step as Kyle suggested in debugging is adding print statements and seeing what the logs show?

Comment (pllim) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:77]: Did you try to use tox and run the exact tox env setting?

Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:77]: I haven't gone quite that far since it takes a while to run, but perhaps it is warranted here... let me give that a go and let it run in the background

Comment (pllim) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:77]: You said on JIRA you were able to reproduce this failure finally, right?

Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:77]: Oops, was that what was implied! Not at all, quite the opposite. Still trying to reproduce. When I ran the tox command, I couldn't get the error to fire either

Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:77]: Actually, @kecnry when you tried running it, did you run it directly via pytest or tox?

Comment (kecnry) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:77]: I never use tox if I can help it.

Comment (pllim) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:77]: Hmm. I guess I'll have a look and see...

File: jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py [hunk 3]
```diff
@@ -41,6 +44,50 @@ def test_plugin(specviz_helper, spectrum1d):
...
+    https://github.com/spacetelescope/jdaviz/pull/1564#discussion_r949427663
+    """
+
+    flux = np.arange(250).reshape((5, 5, 10)) * u.Jy
+    wcs_dict = {"CTYPE1": "RA---TAN", "CTYPE2": "DEC--TAN", "CTYPE3": "WAVE-LOG",
+                "CRVAL1": 205, "CRVAL2": 27, "CRVAL3": 4.622e-7,
+                "CDELT1": -0.0001, "CDELT2": 0.0001, "CDELT3": 8e-11,
+                "CRPIX1": 0, "CRPIX2": 0, "CRPIX3": 0}
+    w = WCS(wcs_dict)
+    large_cube = Spectrum1D(flux=flux, wcs=w)
```
Comment (pllim) [inline jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:61]: Why is this called "large_cube"?

File: jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py (hunk 2 above)
Comment (pllim) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:77]: #1627 somehow has the failing job here passing but the rest fails with that same error... 🤯

File: jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py (hunk 3 above)
Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:61]: The reason is that it's a slightly (larger) modified version of the default spectrum1d cube on our tests. I wanted to make it easier to make a spectral selection for the sake of this pr

Comment (pllim) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:61]: So it is more like "a little larger cube" instead of like "big data cube"?

**Comment by pllim** [comment]
I rewrote your test and it passed on CI, so hopefully it will continue to pass and that wasn't a transient happy accident. 🤞 

See https://github.com/duytnguyendtn/jdaviz/pull/1

The original test case is perhaps too unrealistic to the point of triggering some edge case behavior where sometimes the calculation would give you a number and sometimes a NaN. I also found that even with the new test case, I need to reduce the width to avoid NaN.

File: jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py (hunk 2 above)
Comment (pllim) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:64]: Update: Hopefully CI will pass now.

Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:64]: Green checks! 🎉

File: CHANGES.rst [hunk 4]
```diff
@@ -11,6 +11,8 @@ Cubeviz
 
 - Image viewers now have linked pan/zoom and linked box zoom. [#1596]
 
+- Added ability to select spatial subset collasped spectrum for Line Analysis [#1583]
```
Comment (pllim) [inline CHANGES.rst:14]:
```suggestion
- Added ability to select spatial subset collapsed spectrum for Line Analysis. [#1583]
```

**Review by pllim — APPROVED** [review]
Minor change log typo aside, it did work as advertised while I was investigating the CI failure, so approving. Thanks!

File: jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py (hunk 1 above)
Comment (kecnry) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:108]: I'd still like to see this addressed/discussed before approving/merging!

Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:108]: I was going to address this in my "review phase" now that the CI is passing. I think I've mellowed on my original intent, so I'm happy to switch it back! Let me do that first and I'll rerequest review

Comment (pllim) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:108]: I am not familiar with model fitting, so maybe @rosteen or @javerbukh would have opinions.

Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:108]: Effectively the distinction I tried to make is between selecting the spatial subset itself, or the spectrum generated from the spatial subset. Effectively, it looks identical on the front end, but it's slightly different here in the backend

Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py]: @kecnry I've reverted those changes so they should match; let me know if it's sufficient, or if I missed something!

Comment (kecnry) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py]: I do see your point, but just really want to strive for self-consistency across plugins, especially with attribute names and options since we're looking to expose these directly in #1401.  This still differs from model fittings spatial dropdown (which is the same use-case: acting on the spectrum corresponding to the referenced spatial subset), in that model fitting has "Entire Cube" and this has "Entire Cube Spectrum".  I definitely think they should be the same, and personally would vote for "Entire Cube" and add any clarification to the user in the dropdown hint and/or docs.  But if others prefer being more verbose, then let's also make the change to model fitting.

Comment (duytnguyendtn) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py]: Ok, I changed the default text/labeling as well. I reworked the `hint` to try and make it explicit. How does that look @kecnry ?

**Review by kecnry — APPROVED** [review]
Seems to work well, thanks!

**Comment by duytnguyendtn** [comment]
Oops something broke, one sec

**Comment by duytnguyendtn** [comment]
Thanks so much for the reviews @kecnry @pllim
